This enables the ES8316 audio codec in Rock Pi 4's
DeviceTree (information about it extracted from Radxa's
kernel 4.4 tree)

Unfortunately, it currently causes a crash on startup.
Needs more debugging.

diff -up kernel/arch/arm64/boot/dts/rockchip/rk3399-rock-pi-4.dtsi.omv~ kernel/arch/arm64/boot/dts/rockchip/rk3399-rock-pi-4.dtsi
--- kernel/arch/arm64/boot/dts/rockchip/rk3399-rock-pi-4.dtsi.omv~	2021-01-02 05:54:18.283475880 +0100
+++ kernel/arch/arm64/boot/dts/rockchip/rk3399-rock-pi-4.dtsi	2021-01-02 06:04:58.387809142 +0100
@@ -127,6 +127,45 @@
 		regulator-max-microvolt = <1400000>;
 		vin-supply = <&vcc5v0_sys>;
 	};
+
+	hdmi_codec: hdmi-codec {
+		status = "okay";
+		compatible = "simple-audio-card";
+		simple-audio-card,format = "i2s";
+		simple-audio-card,mclk-fs = <256>;
+		simple-audio-card,name = "HDMI-CODEC";
+
+		simple-audio-card,cpu {
+			sound-dai = <&i2s2>;
+		};
+
+		simple-audio-card,codec {
+			sound-dai = <&hdmi>;
+		};
+	};
+
+	sound_ext_card: sound-ext-card {
+		status = "disabled";
+	};
+
+	spdif-sound {
+		status = "disabled";
+		compatible = "simple-audio-card";
+		simple-audio-card,name = "ROCKCHIP,SPDIF";
+		simple-audio-card,cpu {
+			sound-dai = <&spdif>;
+		};
+		simple-audio-card,codec {
+			sound-dai = <&spdif_out>;
+		};
+	};
+
+	spdif_out: spdif-out {
+		status = "disabled";
+		compatible = "linux,spdif-dit";
+		#sound-dai-cells = <0>;
+	};
+
 };
 
 &cpu_l0 {
@@ -425,6 +485,17 @@
 	i2c-scl-rising-time-ns = <300>;
 	i2c-scl-falling-time-ns = <15>;
 	status = "okay";
+
+	es8316: es8316@11 {
+		#sound-dai-cells = <0>;
+		compatible = "everest,es8316";
+		reg = <0x11>;
+		clocks = <&cru SCLK_I2S_8CH_OUT>;
+		clock-names = "mclk";
+		pinctrl-names = "default";
+		pinctrl-0 = <&i2s_8ch_mclk>;
+	};
+
 };
 
 &i2c3 {
@@ -559,6 +630,28 @@
 			rockchip,pins = <0 RK_PA3 RK_FUNC_GPIO &pcfg_pull_none>;
 		};
 	};
+
+	headphone {
+		hp_det: hp-det {
+			rockchip,pins =
+				<1 0 RK_FUNC_GPIO &pcfg_pull_up>;
+		};
+	};
+
+	i2s0 {
+		i2s0_8ch_bus: i2s0-8ch-bus {
+			rockchip,pins =
+				<3 24 1 &pcfg_pull_none>,
+				<3 25 1 &pcfg_pull_none>,
+				<3 26 1 &pcfg_pull_none>,
+				<3 27 1 &pcfg_pull_none>,
+				<3 31 1 &pcfg_pull_none>;
+		};
+
+		i2s_8ch_mclk: i2s-8ch-mclk {
+			rockchip,pins = <4 0 1 &pcfg_pull_none>;
+		};
+	};
 };
 
 &pwm2 {
